### PR TITLE
Add Web Worker Offloading module

### LINKED ISSUE #176: Partytown support
# Description

I'm opening this issue to discuss how WordPress could take advantage of Partytown to improve its performance.

## What is Partytown?

[Partytown](https://partytown.builder.io/) is a library to execute scripts inside a Web Worker and off the main thread. Its main goal is to increase the site's performance freeing the main thread by offloading non-critical scripts to a Web Worker. It acts as a bridge between the Web Worker and the DOM in a way that is 100% transparent to the executed script.

## How does Partytown work?

Scripts use the type `"text/partytown"` to avoid running in the main thread. Partytown loads those scripts inside a Web Worker that has a proxied version of all the DOM APIs. When the script access a DOM API, the proxy intercepts it and uses one of these two implementations:

1. If Atomics are available, Partytown uses them to communicate synchronously between the main and worker thread.
2. If they are not, Partytown uses synchronous XHR requests intercepted by a Service Worker.

If neither Atomics nor Service Workers are available, Partytown just runs the scripts in the main thread.

It's probably worth noting that when a script executed inside a Partytown Web Worker adds a new script (using APIs like `document.createElement('script')`), that script is also added inside of the Web Worker and not leaked to the main thread.

More information about how it works in [their official docs](https://partytown.builder.io/how-does-partytown-work).

## Partytown tradeoffs

1. DOM operations are slower (and even throttled to reduce work in the main thread), so it should not be used for intensive UI scripts.
2. Partytown fetches all the scripts using `fetch()`, so they need to have CORS headers (or they need to use a proxy).
3. Events handled by scripts that call `event.preventDefault()` will have no effect.

More info about the tradeoffs in [their official docs](https://partytown.builder.io/trade-offs#throttled-dom-operations).

## Restricted access to DOM APIs

Because scripts executed with Partytown are sandboxed and it controls the bridge, site owners can increase the security/control over those scripts by allowing or denying access to the DOM APIs on a per-API basis.

_I don't think Partytown has an API to control this yet, but they have mentioned it several times._

## Partytown Status

Partytown is still a young project. It moved from Alpha to [Beta on 5th Jan 2022](https://www.builder.io/blog/partytown-is-now-in-beta).

# Goals

  

1. Allow plugin developers to execute scripts inside a Partytown Web Worker.
2. Allow site admins to select which scripts they want to execute inside a Partytown Web Worker.

# Proposals

## 1. Add a `worker-partytown` strategy on `wp_enqueue_script`

Following @adamsilverstein proposal of a `strategy` parameter on `wp_enqueue_script` ([#168](https://github.com/WordPress/performance/issues/168)), plugin developers could simply opt-in to execute any script in a Partytown Web Worker by using the `worker-partytown` strategy.

```php
wp_enqueue_script(
  'non-critical-script',
  plugins_url( 'non-critical-script.js', __FILE__ ),
  array(),
  '1.0.0',
  'worker-partytown'
);
```

_There could be other worker strategies as well, so I'd suggest using `worker-partytown` for this one. For example `worker-dom` for [WorkerDOM](https://github.com/ampproject/worker-dom) or simply `worker` for a regular Web Worker._

With this API, plugins could start offering the Web Worker option to their users:

Web Worker option in plugins

We could expose a filter to allow plugins to [configure](https://partytown.builder.io/configuration#vanilla-config) Partytown:

```php
function my_plugin_partytown_config( $config ) {
  $config["debug"] = true;
  return $config;
}
add_filter( 'partytown_configuration', 'my_plugin_partytown_config' );
```

```html
<script>
  partytown = {
    debug: true,
  };
</script>
```

## 2. Partytown module/plugin

Create a module that exposes a user interface to let the site admins configure which services they want to execute inside a Partytown Web Worker.

We could add integrations one by one, making sure Partytown supports those services. If necessary, it should also take care of configuring the [forwarded events](https://partytown.builder.io/forwarding-events).

Integrations could use regular expressions to find the scripts in the HTML and add the `type="text/partytown"` attribute to the script tag. For example, identify scripts with `https://www.google-analytics.com/analytics.js` and/or `ga(...)` for Google Analytics.

```html
<!-- Loaded in Partytown Web Worker -->
<script type="text/partytown">
  (function (i, s, o, g, r, a, m) {
    /* ... */
  })(
    window,
    document,
    "script",
    "https://www.google-analytics.com/analytics.js",
    "ga"
  );
  ga("create", "", "auto");
  ga("send", "pageview");
</script>
```

If this module is successful, we could export its functionality to a Partytown plugin.

## Service Worker integration

We would probably want to integrate it with the [PWA plugin](https://wordpress.org/plugins/pwa/) and only add the Service Worker directly if the PWA plugin is not present.

### PR DESCRIPTION
**Pull request by thelovekesh**
## Summary

Fixes #176

**Please check #271 for previous conversations.**

## Relevant technical choices
- Add Web Worker module which allows running scripts within the worker thread with the help of the Partytown library.
- Add the Partytown package to the dependencies
- Add npm script to copy Partytown lib files to module `assets/js` folder.
    - Currently, Partytown lib files are committed to the module.
    - We can also set up webpack to copy partytown js files in the module build folder when creating the release of the plugin.
    - Partytown library needs to be updated when a new version is released.
- Add tests for the Web Worker Module.
- Add `partytown_configuration` filter which can be used to update partytown [configurations](https://partytown.builder.io/configuration).

## Screencast
https://user-images.githubusercontent.com/54371619/159474078-d88fa7f7-293f-4347-90f4-4ab0d4881240.mp4

## Checklist

- [x] PR has either `[Focus]` or `Infrastructure` label.
- [x] PR has a `[Type]` label.
- [x] PR has a milestone or the `no milestone` label.

### CODE REVIEW

File: .github/workflows/deploy-dotorg.yml
Comment (thelovekesh) [inline .github/workflows/deploy-dotorg.yml]: > https://github.com/WordPress/performance/pull/271#discussion_r993678322

@adamsilverstein Please check I have moved the process of copying partytown files to the release process. 

Since the plugin doesn't have its build system I think it's the best possible place to hook this functionality.

File: modules/javascript/partytown-web-worker/load.php
Comment (adamsilverstein) [inline modules/javascript/partytown-web-worker/load.php]: Rather than using a partytown dependency, can we look for scripts that have a `partytown-worker` strategy (and maybe no dependencies)? 

That way partytown could be enabled with:
`wp_script_add_data( $handle, 'strategy', 'partytown-worker' );`

File: modules/javascript/partytown-web-worker/load.php
Comment (adamsilverstein) [inline modules/javascript/partytown-web-worker/load.php]: adding a filter for each partytown enabled handle seems excessive - it should be a bit cleaner and more efficient to add the filter **once**, then check if the handle is in the list of partytown handles.

File: modules/javascript/partytown-web-worker/load.php
Comment (adamsilverstein) [inline modules/javascript/partytown-web-worker/load.php]: I believe all of the "@ since" annotations can use n.e.x.t to be transformed to the proper version on release.

File: modules/javascript/partytown-web-worker/load.php
Comment (adamsilverstein) [inline modules/javascript/partytown-web-worker/load.php]: Does partytown require https? if so we should check that the site url does as well

**Comment by adamsilverstein** [comment]
@thelovekesh apologies for dropping my attention here: I finally got back to this and left some feedback on the PR; I'd love to help get this over the finish line to land as a module.

**Comment by thelovekesh** [comment]
Thanks @adamsilverstein. I will try to resolve the feedback and update you once done.

File: .github/workflows/deploy-dotorg.yml
Comment (adamsilverstein) [reply .github/workflows/deploy-dotorg.yml]: possibly this could now be added to part of the `npm run build-plugins` command, what do you think?

File: modules/javascript/partytown-web-worker/load.php
Comment (westonruter) [inline modules/javascript/partytown-web-worker/load.php]:
```suggestion
	wp_print_inline_script_tag( sprintf( 'window.partytown = %s;', wp_json_encode( $config ) ) );
```

Comment (westonruter) [reply modules/javascript/partytown-web-worker/load.php]: Or alternatively, this would be even better added as an inline script for the `partytown` script.

File: modules/javascript/partytown-web-worker/load.php
Comment (westonruter) [inline modules/javascript/partytown-web-worker/load.php]:
```suggestion
					$tag = wp_get_script_tag( array( 'type' => 'text/partytown', 'src' => $src, 'id' => "{$handle}-js" ) );
```

File: .github/workflows/deploy-dotorg.yml
Comment (westonruter) [reply .github/workflows/deploy-dotorg.yml]: Having a build process is also needed for Image Loading Optimization. Referenced in #869 pointing to https://github.com/WordPress/performance/pull/876#discussion_r1380553929. We need to use a local copy of web-vitals.js.

File: modules/javascript/partytown-web-worker/load.php
Comment (westonruter) [reply modules/javascript/partytown-web-worker/load.php]: Yes, it would, since service workers require HTTPS. That is, unless the site is `localhost` in which browsers will it like HTTPS. 

Partytown does not require service workers, but they can be easier to deploy. The better approach is using [Atomics](https://partytown.builder.io/atomics), but it has other requirements with COOP/COEP.

File: modules/javascript/partytown-web-worker/load.php
Comment (westonruter) [inline modules/javascript/partytown-web-worker/load.php]: The `wp_print_scripts` action can fire multiple times on the page, such as when someone manually calls `wp_print_scripts( $handles )`. Probably better to do:

```suggestion
add_action( 'wp_print_footer_scripts', 'perflab_web_worker_partytown_worker_scripts' );
```

File: modules/javascript/partytown-web-worker/load.php
Comment (thelovekesh) [reply modules/javascript/partytown-web-worker/load.php]: Updated the logic to hook only single filter.

File: modules/javascript/partytown-web-worker/load.php
Comment (thelovekesh) [reply modules/javascript/partytown-web-worker/load.php]: Updated

File: modules/javascript/partytown-web-worker/load.php
Comment (thelovekesh) [reply modules/javascript/partytown-web-worker/load.php]: Now adding it as an inline script on `partytown` handle.

File: modules/javascript/partytown-web-worker/load.php
Comment (thelovekesh) [reply modules/javascript/partytown-web-worker/load.php]: After removing the dependence from the `wp_print_scripts`, the `script_loader_tag` works. I'm not sure why I initially included this.

File: modules/javascript/partytown-web-worker/load.php
Comment (thelovekesh) [reply modules/javascript/partytown-web-worker/load.php]: While it's feasible to execute this, my inclination is towards considering `partytown` to be included in the `$deps` array as a dependency. 

This preference stems from the fact that when a script loads as `script[type="text/partytown"]`, it necessitates the presence of `partytown`. Thus, referring to it as a direct dependency seems logical.

 Additionally, by adhering to the dependency approach, merely registering the `partytown` script suffices, as WordPress automatically enqueues it if another script depends on it else it will not be enqueued to the front-end.

File: modules/javascript/partytown-web-worker/load.php
Comment (thelovekesh) [reply modules/javascript/partytown-web-worker/load.php]: Humm, that is a good case to consider. In that case, would it be better to disable the module itself? if the site is not using SSL with a notice like:
```txt
The partytown web worker module only works if the site is SSL enabled.
```

File: .github/workflows/deploy-dotorg.yml
Comment (thelovekesh) [reply .github/workflows/deploy-dotorg.yml]: @westonruter @adamsilverstein Added a CLI command called `copy-vendor-libs` in https://github.com/WordPress/performance/pull/556/commits/e6a51cecf3e716b2a263838b970e4b6574bebeb5.

Tasks it will perform:
- Copy vendor files from source to destination based on `copy-vendor-libs` config in `package.json`.
- Copy files from source to destination based on command args i.e. `./bin/plugin/cli.js copy-vendor-libs --source foo/bar/lib --destination modules/module`
- Run after executing the `npm install` command (prepare hook).
- Run with the `build-plugins` command, just before creating the standalone plugin builds.

File: admin/load.php
Comment (thelovekesh) [inline admin/load.php]: Maybe this is not required. Should we include the `partytown` module in the `js-and-css` focus area?

File: modules/javascript/partytown-web-worker/load.php
Comment (westonruter) [inline modules/javascript/partytown-web-worker/load.php]: If the file doesn't exist, shouldn't this do `wp_die()` instead of add an empty string? This file not existing indicates a development issue which wouldn't occur in production.

File: package.json
Comment (westonruter) [inline package.json]: Note: Partytown also has a built-in Copy Task command which perhaps could be reused? https://partytown.builder.io/copy-library-files

File: modules/javascript/partytown-web-worker/load.php
Comment (westonruter) [reply modules/javascript/partytown-web-worker/load.php]: Yes, that makes sense to me, if not `localhost`.

File: package.json
Comment (thelovekesh) [reply package.json]: > Partytown also has a built-in Copy Task command which perhaps could be reused?

If we will be having our build system(most probably we will be having that for the ILO module), that wouldn't be required.

File: modules/javascript/partytown-web-worker/load.php
Comment (thelovekesh) [reply modules/javascript/partytown-web-worker/load.php]: That makes sense. and maybe with a notice?

Comment (westonruter) [reply modules/javascript/partytown-web-worker/load.php]: A notice in `wp_die()`? Ideally we wouldn't include the `file_exists()` check in the build of the plugin, since we wouldn't need to keep checking for it all the time as it would be guaranteed to be there. The condition could be stripped out for distribution. This is something that could be done in [`build-plugins`](https://github.com/WordPress/performance/blob/0e7f28516e452e107fbaf66a3a548c274f230878/bin/plugin/commands/build-plugins.js#L67-L94), although that is for standalone _plugins_ and not modules. In any case, I don't think that this should necessarily hold up this PR. There should just be some error communicated that the user hasn't done a build.

File: package.json
Comment (westonruter) [reply package.json]: Yeah, you're probably right.

File: modules/javascript/partytown-web-worker/load.php
Comment (thelovekesh) [reply modules/javascript/partytown-web-worker/load.php]: Added in https://github.com/WordPress/performance/pull/556/commits/4191312440cca374e0c377f8b277acacc887ab5e

File: modules/javascript/partytown-web-worker/load.php
Comment (thelovekesh) [reply modules/javascript/partytown-web-worker/load.php]: Agree since FS checks are costly. To address this:
- Add logic to not load Partytown module at all if vendor files does not exists and add admin notice - https://github.com/WordPress/performance/pull/556/commits/8c76f52159f0b386dc4e026e9dc2080f3eae99b9
- Remove file exists check for Partytown library - https://github.com/WordPress/performance/pull/556/commits/16a466c7f5181fda8f3ff3af2139ffe821fc8b6b

File: modules/javascript/partytown-web-worker/can-load.php
Comment (westonruter) [inline modules/javascript/partytown-web-worker/can-load.php]: Why are these pluggable functions?

File: modules/javascript/partytown-web-worker/can-load.php
Comment (westonruter) [inline modules/javascript/partytown-web-worker/can-load.php]: Service workers normally don't work on HTTPS, but the exception is on `localhost` for development. Also passing `url` is just a wrapper for `home_url()`:

```suggestion
			( wp_parse_url( home_url(), PHP_URL_SCHEME ) !== 'https' )
			&&
			'localhost' !== wp_parse_url( home_url(), PHP_URL_HOST )
```

File: modules/javascript/partytown-web-worker/can-load.php
Comment (westonruter) [inline modules/javascript/partytown-web-worker/can-load.php]: Passing `wpurl` is just an alias for `site_url()`:

```suggestion
			( wp_parse_url( site_url(), PHP_URL_SCHEME ) !== 'https' )
```

File: admin/load.php
Comment (westonruter) [reply admin/load.php]: Yeah, let's include as part of `js-and-css`

File: modules/javascript/partytown-web-worker/can-load.php
Comment (westonruter) [inline modules/javascript/partytown-web-worker/can-load.php]: Nevertheless, I think probably the `is_ssl()` and `localhost` checks would be sufficient?

File: modules/javascript/partytown-web-worker/load.php
Comment (adamsilverstein) [reply modules/javascript/partytown-web-worker/load.php]: That makes sense, however I can also see the use case for the `wp_script_add_data` approach. For example, an optimization plugin could use this to specify it wants a specific script loaded with partytown. I'm not sure you could add a dependency dynamically like that (I have to look). 

This is fine as is for now, I will check if maybe we can make it work both ways.

**Comment by adamsilverstein** [comment]
Looks really good, I'm going to try to test this out a bit locally.

**Comment by thelovekesh** [comment]
Note: Partytown lib assets can be copied using the Webpack which landed recently in the repo.
See https://github.com/WordPress/performance/pull/953

File: module-i18n.php
Comment (adamsilverstein) [inline module-i18n.php]:
```suggestion
	_x( 'Add support for running JavaScript in a separate Web Worker with the help of the Partytown library.', 'module description', 'performance-lab' ),
```

Comment (thelovekesh) [reply module-i18n.php]: I think it's an auto-generated file. Better to update this in the plugin header?

Comment (westonruter) [reply module-i18n.php]: Wouldn't it be in both? Curious why `module-i18n.php` is committed to git in the first place. Seems like it should be part of the build process only? cc @mukeshpanchal27

File: modules/javascript/partytown-web-worker/can-load.php
Comment (thelovekesh) [reply modules/javascript/partytown-web-worker/can-load.php]: Fixed in https://github.com/WordPress/performance/pull/556/commits/ac2c9d0d5b900969699c35979a5b4afabb2aae97

File: modules/javascript/partytown-web-worker/can-load.php
Comment (thelovekesh) [reply modules/javascript/partytown-web-worker/can-load.php]: Updated in https://github.com/WordPress/performance/pull/556/commits/ac2c9d0d5b900969699c35979a5b4afabb2aae97

File: .github/workflows/deploy-dotorg.yml
Comment (thelovekesh) [reply .github/workflows/deploy-dotorg.yml]: We have removed the above logic and now using Webpack to copy the assets. See https://github.com/WordPress/performance/pull/556/commits/5dfcfdfee96f090cd18889dbc6bc9b3a68d875ac and https://github.com/WordPress/performance/pull/953

File: admin/load.php
Comment (thelovekesh) [reply admin/load.php]: Done in https://github.com/WordPress/performance/pull/556/commits/335daaa1ace427486f7b2023963ffeda72f6befc

File: .github/CODEOWNERS
Comment (swissspidy) [inline .github/CODEOWNERS:52]: Note that these would only work when given write access to the repo

File: modules/js-and-css/partytown-web-worker/load.php
Comment (swissspidy) [inline modules/js-and-css/partytown-web-worker/load.php:3]: Do we really want to call the module Partytown?

IMHO usage of the Partytown library is just an implementation detail.

Comment (westonruter) [reply modules/js-and-css/partytown-web-worker/load.php:3]: Agreed. A less-specific name would be "Web Worker Offloading". I asked Bard for some alternative names, and it suggested "Threadshift". I like that, but it is trademarked.

**Comment by westonruter** [comment]
Since we're currently in the middle of re-plumbing for [standalone plugins](https://github.com/WordPress/performance/issues?q=is%3Aissue+is%3Aopen+label%3A%22Creating+standalone+plugins%22), I'm going to re-base this branch to a ~`feature/partytown`~ `feature/web-worker-offloading` branch. I don't think this feature should be released as a module but rather as a standalone plugin, right? cc @mukeshpanchal27 @joemcgill

File: modules/js-and-css/partytown-web-worker/load.php
Comment (thelovekesh) [reply modules/js-and-css/partytown-web-worker/load.php:3]: That makes sense. If you have a certain name in mind, do let me know, and I will edit it.

Comment (westonruter) [reply modules/js-and-css/partytown-web-worker/load.php:3]: How about Web Worker Offloading for now?

Comment (thelovekesh) [reply modules/js-and-css/partytown-web-worker/load.php:3]: SGTM, some more names:
- Scripts offloading/offloader (in the description add a reference to Web Worker and/or Partytown).
- Third-party JS offloader
- Non-critical JS offloader

I will defer name selection to you.

Comment (westonruter) [reply modules/js-and-css/partytown-web-worker/load.php:3]: Those are some good options but I like Web Worker Offloading.

The JS being offloaded may not be third-party. It may also be critical, but just offloaded so it can run in another thread.

File: .gitignore
Comment (westonruter) [inline .gitignore:37]: This will necessitate introducing a `.distignore` file to replace the `.gitattributes` file as explained in https://github.com/WordPress/performance/issues/893#issuecomment-1947398713.

File: module-i18n.php
Comment (westonruter) [inline module-i18n.php:12]:
```suggestion
	_x( 'Web Worker Offloading', 'module name', 'performance-lab' ),
```

File: modules/js-and-css/partytown-web-worker/load.php
Comment (westonruter) [reply modules/js-and-css/partytown-web-worker/load.php:3]:
```suggestion
 * Module Name: Web Worker Offloading
```

File: tests/modules/js-and-css/partytown-web-worker/partytown-web-worker-test.php
Comment (westonruter) [inline tests/modules/js-and-css/partytown-web-worker/partytown-web-worker-test.php:3]:
```suggestion
 * Test cases for Web Worker Offloading module.
```

File: modules/js-and-css/partytown-web-worker/load.php
Comment (westonruter) [inline modules/js-and-css/partytown-web-worker/load.php:88]: Correct me if I'm wrong, but let's say you have an existing script in a theme like:

```php
wp_enqueue_script( 'foo', 'https://example.com/foo.js' );
```

In order to opt-in this script to use Partytown, you'd have to do this if you wanted to opt-in from a separate plugin:

```php
wp_scripts()->registered['foo']->deps[] = 'partytown';
```

Is that right?

**Review by westonruter — COMMENTED** [review]
Since this would certainly get released as a standalone plugin on Wordpress.org, it seems we'd need to have the plugin actually do something without needing to code. Otherwise, I understand that the plugin review team would reject the plugin as a "framework plugin". I recall there was some discussion of what this could do by default, without there being a clear answer. Would there need to be a UI to select scripts to opt-in? Humm.

**Comment by adamsilverstein** [comment]
> Since this would certainly get released as a standalone plugin on Wordpress.org, it seems we'd need to have the plugin actually do something without needing to code 

We could have it automatically apply to analytics scripts where it is expected to work well - although those may not be enqueued with wp_scripts.

File: modules/js-and-css/partytown-web-worker/load.php
Comment (thelovekesh) [reply modules/js-and-css/partytown-web-worker/load.php:88]: Yup, that should do the trick.

**Comment by thelovekesh** [comment]
> We could have it automatically apply to analytics scripts where it is expected to work well - although those may not be enqueued with wp_scripts.

If the user wishes to offload these [common scripts](https://partytown.builder.io/common-services), it would be preferable if there was an opt-in form on some sort of options page. Alternatively, the user may utilize this as a framework to offload scripts.

There are two scenarios I can think of if the user opts in for script auto-offloading:
- They should be enqueued with wp_scripts, and we could use something like `wp_scripts()->registered['foo']->deps[] = 'partytown';` 
- Alternatively, if any script tags are inline scripts added to the page's head or footer, we will need to parse every script tag on the page and offload the above-mentioned common scripts if available before the output buffer.

**Comment by westonruter** [comment]
Just noting this can be updated to use the new standalone plugin structure.

**Review by westonruter — APPROVED** [review]
Approving to continue work in the feature branch

**Review by adamsilverstein — APPROVED** [review]
Looking good! Lets keep iterating and testing in the branch.

**Pull request merged**
